sklearn/utils/validation.py

```diff
@@ -97,16 +98,24 @@ def _assert_all_finite(
 
     if _get_config()["assume_finite"]:
         return
+
     X = np.asanyarray(X)
     # First try an O(n) time, O(1) space solution for the common case that
     # everything is finite; fall back to O(n) space np.isfinite to prevent
     # false positives from overflow in sum method.
     is_float = X.dtype.kind in "fc"
-    with np.errstate(over="ignore"):
-        first_pass_isfinite = is_float and np.isfinite(np.sum(X))
-    if first_pass_isfinite:
-        pass
-    elif is_float:
+    # Cython implementation doesn't support FP16 or complex numbers
+    # size > 5000 is a heuristic for when the python implementation is faster
+    use_cython = not (X.dtype.kind == "c" or X.dtype == np.float16) and X.size > 5000
+    first_pass_isfinite = False
+    if is_float:
+        if use_cython:
+            first_pass_isfinite = cy_isfinite(X, allow_nan=allow_nan)
+        else:
+            with np.errstate(over="ignore"):
+                first_pass_isfinite = np.isfinite(np.sum(X))
```

Comment:
Could this be done in both the and Cython and Python cases? IOW only using `cy_isfinite` after `np.isfinite` failed